components/function-controller/internal/git/git.go

```diff
@@ -0,0 +1,74 @@
+package git
+
+import (
+	"fmt"
+
+	gogit "github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/transport"
+	"github.com/pkg/errors"
+)
+
+const refsHeadsPrefix = "refs/heads/%s"
+
+//go:generate mockery -name=Client -output=automock -outpkg=automock -case=underscore
+type Client interface {
+	ListRefs(repoUrl string, auth transport.AuthMethod) ([]*plumbing.Reference, error)
+	PlainClone(path string, isBare bool, options *gogit.CloneOptions) (*gogit.Repository, error)
+}
+
+type Git struct {
+	client Client
+}
+
+func New() *Git {
+	return &Git{client: newClient()}
+}
+
+func (g *Git) LastCommit(repoUrl, branch string, auth transport.AuthMethod) (commitHash string, err error) {
```

Comment:
Also in other places

```suggestion
func (g *Git) LastCommit(repoUrl, branch string, auth transport.AuthMethod) (string, error) {
```